### PR TITLE
Improve sound stuff

### PR DESCRIPTION
**Pull request by Blueyescat**
### Description
* Fixes the issues with play sound effect and improves its code
* Adds more description and examples to the play sound effect
* Adds multiple sounds support to the play sound effect
* Adds multiple locations support to the play sound effect
* Adds ability to specify sound category in the play sound effect
* Adds ability to playing sounds to players without specifying a location as suggested by @TheLimeGlass in #1921.
* Adds stop sound effect
* Adds sound category enum type, used in play and stop sound effects
* Adds requiredPlugins method to ClassInfo because i needed it for the sound category type
* Fixes inconsistent empty line indentations in BukkitClasses, cleans up its code

---
**Target Minecraft Versions:** any, 1.10.2+ for Stop Sound, 1.11+ for sound categories
**Requirements:** none
**Related Issues:** #1914 and  #1949

### CODE REVIEW

**Comment by TheLimeGlass** [comment]
I have some minor suggestions though.

The SOUND_CATEGORIES_EXIST field doesn't need to be static.
The Local in the toUpperCase isn't needed, and does Skript check if the event is null in the toString of expressions, if so the check if the event is null is unneeded.

Maybe another third pattern to support defining the sound category to use?

Looks like a pretty good commit. Been a long time coming.

**Comment by Blueyescat** [comment]
@TheLimeGlass 

> The SOUND_CATEGORIES_EXIST field doesn't need to be static.

Might be correct, i used this way in other PRs too..

> The Local in the toUpperCase isn't needed

It is needed, https://github.com/SkriptLang/Skript/pull/1842

> does Skript check if the event is null in the toString of expressions, if so the check if the event is null is unneeded.

It is checked in some effects but i have no idea whether it is needed. At least the old code had it and i didn't remove.

> Maybe another third pattern to support defining the sound category to use?

I don't see any benefits.

**Comment by TheLimeGlass** [comment]
I know you may not see a benefit in it, but some user out there may fine use in it.

In my opinion it is always best to add support for something.

I always make over the configurations in my plugins to allow for the user to decide on what they feel they need.

**Comment by Blueyescat** [comment]
Do the users know what are they

**Comment by TheLimeGlass** [comment]
Some will, you can put an explaination in the description or in the description of the Class info for sound categories.

Also adding support for the `%sounds/strings%` would be a good addition as said above.

**Comment by Blueyescat** [comment]
From what i see sound categories are just to change the sound volume. **master** is always the highest. I can't even find any info about sound categories in the Minecraft Wiki.

And no i don't think to add sound enums in this PR

**Comment by TheLimeGlass** [comment]
Oh you don't know what SoundCategory is, I thought you were asking if users knew.

SoundCategory allows you to define the category a sound is played in.

You know when you go into the sound settings in Minecraft, there are a bunch of sliders. You can define which category to send the sound to.

**Comment by Blueyescat** [comment]
Oh i totally forgot about the settings lol. Then they can be definitely useful, yeah i can add support for them with an enum type. Do you have a syntax suggestion?

**Comment by Pikachu920** [comment]
>The SOUND_CATEGORIES_EXIST field doesn't need to be static.

it should be static because it's a constant independent of any instance

**Comment by TheLimeGlass** [comment]
It doesn't need to be static, Skript will already turn the effect into an instance of an Effect. There is no reason for it to be static. It's only static so it can be used in the static registration, which is poor, but grandfathered in.

**Comment by Nicofisi** [comment]
> It doesn't need to be static, Skript will already turn the effect into an instance of an Effect, so it's actually not independent. There is no reason for it to be static. It's only static so it can be used in the static registration, which is poor, but grandfathered in.

I don't know what you're talking about, but from my point of view, the only reason it's static, is so that the check doesn't need to run every time the effect is found somewhere in a script, and so that the value doesn't need to be recomputed and stored multiple times in the RAM. 

I don't see any downsides of this field being static, either. Do you expect the `org.bukkit.SoundCategory` class to somehow reappear and be available later?

**Comment by TheLimeGlass** [comment]
> 
> 
> > It doesn't need to be static, Skript will already turn the effect into an instance of an Effect, so it's actually not independent. There is no reason for it to be static. It's only static so it can be used in the static registration, which is poor, but grandfathered in.
> 
> I don't know what you're talking about, but from my point of view, the only reason it's static, is so that the check doesn't need to run every time the effect is found somewhere in a script, and so that the value doesn't need to be recomputed and stored multiple times in the RAM.
> 
> I don't see any downsides of this field being static, either. Do you expect the `org.bukkit.SoundCategory` class to somehow reappear and be available later?

It doesn't need to be static because just like you said, it's only being used in the actual calling method.

**Comment by Nicofisi** [comment]
But making it static makes it so 

> > > It doesn't need to be static, Skript will already turn the effect into an instance of an Effect, so it's actually not independent. There is no reason for it to be static. It's only static so it can be used in the static registration, which is poor, but grandfathered in.
> > 
> > 
> > I don't know what you're talking about, but from my point of view, the only reason it's static, is so that the check doesn't need to run every time the effect is found somewhere in a script, and so that the value doesn't need to be recomputed and stored multiple times in the RAM.
> > I don't see any downsides of this field being static, either. Do you expect the `org.bukkit.SoundCategory` class to somehow reappear and be available later?
> 
> It doesn't need to be static because just like you said, it's only being used in the actual calling method.

But making it static makes it so the value only needs to be computed once

**Comment by TheLimeGlass** [comment]
> 
> 
> But making it static makes it so
> 
> > > > It doesn't need to be static, Skript will already turn the effect into an instance of an Effect, so it's actually not independent. There is no reason for it to be static. It's only static so it can be used in the static registration, which is poor, but grandfathered in.
> > > 
> > > 
> > > I don't know what you're talking about, but from my point of view, the only reason it's static, is so that the check doesn't need to run every time the effect is found somewhere in a script, and so that the value doesn't need to be recomputed and stored multiple times in the RAM.
> > > I don't see any downsides of this field being static, either. Do you expect the `org.bukkit.SoundCategory` class to somehow reappear and be available later?
> > 
> > 
> > It doesn't need to be static because just like you said, it's only being used in the actual calling method.
> 
> But making it static makes it so the value only needs to be computed once

But Skript saves all the initiated Effect objects in https://github.com/SkriptLang/Skript/blob/master/src/main/java/ch/njol/skript/Skript.java#L1160

Static would not change anything.

"Each call to "new" would create a new instance of the class and the fields"

**Comment by Pikachu920** [comment]
>But Skript saves all the initiated Effect objects in https://github.com/SkriptLang/Skript/blob/master/src/main/java/ch/njol/skript/Skript.java#L1160

The line you posted stores `SyntaxElementInfo`s for each registered effect class, not all instances of Effect. Making it not static means each instance would have it's own `SOUND_CATEGORIES_EXIST` which would result in increased memory usage and redundant computation.

**Comment by Nicofisi** [comment]
@TheLimeGlass the line you linked stores a collection of `SyntaxElementInfo`s, not `Effect`s. Wouldn't storing all initialized effect objects for use everywhere mean that effects like [EffHealth](https://github.com/SkriptLang/Skript/blob/master/src/main/java/ch/njol/skript/effects/EffHealth.java) that have two different behaviors (either heal or damage) could only have one of them, per server?

**Comment by TheLimeGlass** [comment]
An effect is an SyntaxElementInfo. It gets stored when Skript loads.

**Comment by Nicofisi** [comment]
Aren't you a little lost? Neither `SyntaxElementInfo`s contain any instances of `Effect`s, nor `Effect`s/`Statement`s are `SyntaxElementInfo`s
https://github.com/SkriptLang/Skript/blob/master/src/main/java/ch/njol/skript/lang/SyntaxElementInfo.java
https://github.com/SkriptLang/Skript/blob/master/src/main/java/ch/njol/skript/lang/Effect.java
https://github.com/SkriptLang/Skript/blob/master/src/main/java/ch/njol/skript/lang/Statement.java

**Comment by TheLimeGlass** [comment]
Discussed in the dev-zone on Discord.

If the class is being created into a new instance everytime it still gets processed that same as if it wasn't static. The static fields and any other declared fields get created per each new instance.

The static field would be best if it doesn't exist, but doesn't matter that much since it unfortunately has to be used within the static block to register.

**Comment by Pikachu920** [comment]
>The static fields and any other declared fields get created per each new instance

I think that is the source of confusion here...

Only instance fields get initialized when a class is instantiated, static fields initialize when the class is first linked (which is typically done immediately after loading). You can verify behavior this in the Java spec [12.3](https://docs.oracle.com/javase/specs/jls/se7/html/jls-12.html#jls-12.3) and [12.3.2](https://docs.oracle.com/javase/specs/jls/se7/html/jls-12.html#jls-12.3.2)

**Comment by bensku** [comment]
When checking for existence of a particular Bukkit API, store the result in a static final variable. This is semantically correct, and is also easiest to optimize for JIT compiler.

File: src/main/resources/lang/english.lang [hunk 1]
```diff
@@ -1379,6 +1379,19 @@ resource pack states:
 	failed_download: download fail, fail, failed to download, failed
 	successfully_loaded: successfully load, successfully install, success, successfully loaded, successfully installed
 
+# -- Sound Categories --
+sound categories:
+	ambient: ambient, environment
```
[the remark this answers is not in the document]
Comment (Blueyescat) [reply src/main/resources/lang/english.lang]: it doesn't

[the remark this answers is not in the document]
Comment (Blueyescat) [reply src/main/resources/lang/english.lang]: Found some issues and added the suffix.

**Comment by bensku** [comment]
Would be good to have this for 2.4. @FranKusmiruk please merge if you still approve this PR.

**Pull request merged**